File: cses/08_string_algorithms/2104_pattern-positions/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    s = data[0]                     # xâu s dạng bytes; lặp qua trả về các số int
    # data[1] là k; các pattern nằm ở phần còn lại.
    patterns = data[2:]

    n = len(s)
    max_states = 2 * n + 5

    # Các mảng song song mô tả từng trạng thái của Suffix Automaton (SAM).
    trans = [None] * max_states     # danh sách dict: byte -> trạng thái
    link = [0] * max_states         # suffix link
    length = [0] * max_states       # độ dài xâu dài nhất của lớp tương đương
    firstpos = [0] * max_states     # vị trí kết thúc (0-indexed) sớm nhất

    # Trạng thái khởi đầu.
    trans[0] = {}
    link[0] = -1
    length[0] = 0
    sz = 1
    last = 0

    # Xây SAM online: thêm lần lượt từng ký tự của s.
    for pos in range(n):
        c = s[pos]
        # Tạo trạng thái mới cur cho tiền tố hiện tại; firstpos[cur] = pos là vị
        # trí kết thúc lần xuất hiện đầu tiên.
        cur = sz
        sz += 1
        trans[cur] = {}
        length[cur] = length[last] + 1
        firstpos[cur] = length[cur] - 1          # == pos
        # Nối transition dọc theo chuỗi suffix link của last.
        p = last
        while p != -1 and c not in trans[p]:
            trans[p][c] = cur
            p = link[p]
        if p == -1:
            link[cur] = 0
        else:
            q = trans[p][c]
            if length[p] + 1 == length[q]:
                link[cur] = q
            else:
                # Tách (clone) q; clone kế thừa firstpos[q] vì endpos chỉ mở rộng
                # thêm vị trí mới (lớn hơn) nên giá trị nhỏ nhất không đổi.
                clone = sz
                sz += 1
                length[clone] = length[p] + 1
                trans[clone] = dict(trans[q])
                link[clone] = link[q]
                firstpos[clone] = firstpos[q]
                while p != -1 and trans[p].get(c) == q:
                    trans[p][c] = clone
                    p = link[p]
                link[q] = clone
                link[cur] = clone
        last = cur

    # Trả lời truy vấn. Nội địa hoá biến để tăng tốc.
    res = []
    append = res.append
    tr = trans
    fp = firstpos
    for pat in patterns:
        # Đi từng ký tự của pattern trong SAM.
        v = 0
        ok = True
        for c in pat:
            nxt = tr[v].get(c)
            if nxt is None:         # thiếu transition -> không phải xâu con
                ok = False
                break
            v = nxt
        if ok:
            # Đổi vị trí kết thúc sớm nhất sang vị trí bắt đầu (1-indexed).
            append(fp[v] - len(pat) + 2)
        else:
            append(-1)

    sys.stdout.write('\n'.join(map(str, res)))
    sys.stdout.write('\n')


main()

File: cses/08_string_algorithms/2104_pattern-positions/solution.cpp
#include <array>
#include <iostream>
#include <string>
#include <vector>

// Một trạng thái của Suffix Automaton (SAM).
struct State {
    int length = 0;          // độ dài xâu dài nhất thuộc lớp tương đương này
    int link = -1;           // suffix link
    int first_position = 0;  // vị trí kết thúc (0-indexed) sớm nhất -> firstpos
    std::array<int, 26> next{};  // transition theo 26 ký tự

    State() {
        next.fill(-1);
    }
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string text;
    if (!(std::cin >> text)) {
        return 0;
    }

    // Xây SAM của xâu s theo từng ký tự (thuật toán online chuẩn CP-algorithms).
    std::vector<State> automaton;
    automaton.reserve(2 * text.size());
    automaton.emplace_back();  // trạng thái khởi đầu
    int last = 0;

    for (int position = 0; position < static_cast<int>(text.size()); ++position) {
        const int letter = text[position] - 'a';

        // Tạo trạng thái mới cur đại diện cho toàn bộ tiền tố hiện tại.
        // firstpos[cur] = position: vị trí kết thúc lần xuất hiện đầu tiên.
        const int current = static_cast<int>(automaton.size());
        automaton.emplace_back();
        automaton[current].length = automaton[last].length + 1;
        automaton[current].first_position = position;

        // Nối transition từ chuỗi suffix link của last tới cur.
        int state = last;
        while (state != -1 && automaton[state].next[letter] == -1) {
            automaton[state].next[letter] = current;
            state = automaton[state].link;
        }

        if (state == -1) {
            automaton[current].link = 0;
        } else {
            const int target = automaton[state].next[letter];
            if (automaton[state].length + 1 == automaton[target].length) {
                // target liền kề, gán suffix link trực tiếp.
                automaton[current].link = target;
            } else {
                // Tách (clone) target; clone kế thừa firstpos[target] vì endpos
                // chỉ mở rộng thêm vị trí mới (lớn hơn) nên min không đổi.
                const int clone = static_cast<int>(automaton.size());
                automaton.push_back(automaton[target]);
                automaton[clone].length = automaton[state].length + 1;

                while (state != -1 && automaton[state].next[letter] == target) {
                    automaton[state].next[letter] = clone;
                    state = automaton[state].link;
                }
                automaton[target].link = clone;
                automaton[current].link = clone;
            }
        }
        last = current;
    }

    // Trả lời truy vấn: đi từng ký tự của pattern trong SAM.
    int pattern_count = 0;
    std::cin >> pattern_count;
    while (pattern_count-- > 0) {
        std::string pattern;
        std::cin >> pattern;
        int state = 0;
        for (const char character : pattern) {
            state = automaton[state].next[character - 'a'];
            if (state == -1) {  // thiếu transition -> pattern không phải xâu con
                break;
            }
        }
        if (state == -1) {
            std::cout << -1 << '\n';
        } else {
            // Đổi vị trí kết thúc sớm nhất sang vị trí bắt đầu (1-indexed).
            std::cout << automaton[state].first_position - static_cast<int>(pattern.size()) + 2 << '\n';
        }
    }

    return 0;
}
